cmd/agent/app/throttling/credit_store.go

```diff
@@ -0,0 +1,261 @@
+}
+
+func newCreditAccruer(serviceName string, options CreditAccruerOptions, timeNow func() time.Time) *creditAccruer {
+	defaultRateLimiter := newTokenBucket(options, timeNow)
+	perOperationRateLimiter := NewOverrideMap(options.MaxOperations, defaultRateLimiter)
+	return &creditAccruer{
+		options:                 options,
+		serviceName:             serviceName,
+		updateTime:              timeNow(),
+		timeNow:                 timeNow,
+		perOperationRateLimiter: perOperationRateLimiter,
+	}
+}
+
+// Withdraw credits from the creditAccruer for the given operationName with an
+// upper limit of maxWithdrawal.
+func (accruer *creditAccruer) Withdraw(operationName string, maxWithdrawal float64) float64 {
+	var rateLimiter *tokenBucket
+	if accruer.perOperationRateLimiter.Has(operationName) ||
+		accruer.perOperationRateLimiter.IsFull() {
+		rateLimiter = accruer.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(accruer.options, accruer.timeNow)
+		accruer.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// CreditStore manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type CreditStore interface {
+	io.Closer
+	// Withdraw deducts as many credits from the service on behalf of the client
+	// without exceeding MaxClientBalance. The client is identified by a unique
+	// clientID string. The credits are deducted from the specific operation on
+	// the service if present, otherwise from the default credit pool.
+	Withdraw(clientID string, serviceName string, operationName string) float64
+	// Spend credits already allocated to this client on a given operation.
+	Spend(clientID string, operationName string, credits float64) bool
+}
+
+type creditStore struct {
+	sync.Mutex
+	creditAccruers map[string]*creditAccruer
+	clients        map[string]*client
+	options        CreditAccruerOptions
+	ttl            time.Duration
+	timeNow        func() time.Time // For testing
+	ticker         *time.Ticker
+	done           chan bool
+	waitGroup      sync.WaitGroup
+}
+
+// NewCreditStore creates a new creditStore and returns it to the caller.
+// options should be the default values passed to new creditAccruers.
+// ttl should be the TTL for all creditAccruers and clients.
+func NewCreditStore(options CreditAccruerOptions, ttl time.Duration) CreditStore {
+	s := &creditStore{
+		creditAccruers: map[string]*creditAccruer{},
+		clients:        map[string]*client{},
+		options:        options,
+		ttl:            ttl,
+	}
+	s.timeNow = time.Now
+	s.ticker = time.NewTicker(10 * time.Second)
+	s.done = make(chan bool)
+	s.waitGroup.Add(1)
+	go func() {
+		defer s.waitGroup.Done()
```

Comment:
ticket should be defined here, and defer-closed. The top level struct doesn't need to know about it.

Reply:
Sure. Out of curiosity, is there a better approach here? A `Ticker`, `channel`, and `WaitGroup` seems like overkill for such a common operation. Even `Mutex` and `ConditionVariable` looks easier to implement.

Reply:
did you need the wait group?  All it does is block the main function until the go routine is running. Some times its useful, not sure if it is in this case. The struct certainly doesn't need to know about the wait group, it can be just a local var.

Re ticker/channel, given that this is a timer loop I don't know if another way if implementing it. And since you're already using select for the ticker channel, the `done` channel makes perfect sense.

Reply:
I guess it isn't strictly necessary but guarantees `Close` is actually finishes before proceeding. Can see this being useful in general, if not for this specific case. I guess this is more of a question for the Go language developers. After a bit of research, it seems the Go API for condition variables is intentionally limited to avoid my alternative approach.